File: src/models/fallback.ts
/**
 * 模型回退机制
 * 实现自动切换到备用模型
 */

import type { FallbackConfig, ModelSwitchEvent } from './types.js';
import { ModelFallbackError } from './types.js';
import { modelConfig } from './config.js';

/**
 * 可重试的错误类型
 */
const DEFAULT_RETRYABLE_ERRORS = [
  'overloaded_error',
  'rate_limit_error',
  'api_error',
  'timeout',
  'ECONNRESET',
  'ETIMEDOUT',
  'ENOTFOUND',
  'capacity_exceeded',
  'model_unavailable',
];

/**
 * 模型回退管理器
 */
export class ModelFallback {
  private config: FallbackConfig;
  private switchHistory: ModelSwitchEvent[] = [];
  private listeners: Array<(event: ModelSwitchEvent) => void> = [];

  constructor(config?: Partial<FallbackConfig>) {
    this.config = {
      primaryModel: config?.primaryModel || modelConfig.getDefaultModel(),
      fallbackModel: config?.fallbackModel,
      retryableErrors: config?.retryableErrors || DEFAULT_RETRYABLE_ERRORS,
      maxRetries: config?.maxRetries ?? 3,
      retryDelayMs: config?.retryDelayMs ?? 1000,
      exponentialBackoff: config?.exponentialBackoff ?? true,
    };
  }

  /**
   * 设置主模型
   */
  setPrimaryModel(model: string): void {
    if (model === this.config.fallbackModel) {
      throw new Error('Primary model cannot be the same as fallback model');
    }
    this.config.primaryModel = modelConfig.resolveAlias(model);
  }

  /**
   * 设置回退模型
   */
  setFallbackModel(model: string | undefined): void {
    if (model === this.config.primaryModel) {
      throw new Error('Fallback model cannot be the same as primary model');
    }
    this.config.fallbackModel = model ? modelConfig.resolveAlias(model) : undefined;
  }

  /**
   * 获取当前配置
   */
  getConfig(): FallbackConfig {
    return { ...this.config };
  }

  /**
   * 检查错误是否可重试
   */
  isRetryable(error: Error | unknown): boolean {
    if (!(error instanceof Error)) {
      return false;
    }

    const errorMessage = error.message.toLowerCase();
    const errorType = (error as any).type || (error as any).code || '';

    return this.config.retryableErrors.some(
      (retryable) =>
        errorMessage.includes(retryable.toLowerCase()) ||
        errorType.toLowerCase().includes(retryable.toLowerCase())
    );
  }

  /**
   * 计算重试延迟
   */
  private getRetryDelay(attempt: number): number {
    if (this.config.exponentialBackoff) {
      return this.config.retryDelayMs * Math.pow(2, attempt);
    }
    return this.config.retryDelayMs;
  }

  /**
   * 等待
   */
  private sleep(ms: number): Promise<void> {
    return new Promise((resolve) => setTimeout(resolve, ms));
  }

  /**
   * 执行带回退的操作
   */
  async executeWithFallback<T>(
    operation: (model: string) => Promise<T>,
    options?: {
      onRetry?: (model: string, attempt: number, error: Error) => void;
      onFallback?: (fromModel: string, toModel: string, error: Error) => void;
    }
  ): Promise<{ result: T; model: string; retries: number; usedFallback: boolean }> {
    const { primaryModel, fallbackModel, maxRetries } = this.config;
    let lastError: Error | null = null;
    let retryCount = 0;
    let currentModel = primaryModel;
    let usedFallback = false;

    // 尝试主模型
    for (let attempt = 0; attempt <= maxRetries; attempt++) {
      try {
        const result = await operation(currentModel);
        return {
          result,
          model: currentModel,
          retries: retryCount,
          usedFallback,
        };
      } catch (error) {
        lastError = error instanceof Error ? error : new Error(String(error));
        retryCount++;

        if (this.isRetryable(error) && attempt < maxRetries) {
          const delay = this.getRetryDelay(attempt);

          if (options?.onRetry) {
            options.onRetry(currentModel, attempt + 1, lastError);
          }

          console.warn(
            `API error (${lastError.message}), retrying in ${delay}ms... ` +
            `(attempt ${attempt + 1}/${maxRetries})`
          );

          await this.sleep(delay);
          continue;
        }

        // 主模型失败，尝试回退模型
        if (currentModel === primaryModel && fallbackModel && this.isRetryable(error)) {
          currentModel = fallbackModel;
          usedFallback = true;

          if (options?.onFallback) {
            options.onFallback(primaryModel, fallbackModel, lastError);
          }

          // 记录切换事件
          this.recordSwitch({
            fromModel: primaryModel,
            toModel: fallbackModel,
            reason: 'fallback',
            timestamp: Date.now(),
            error: lastError.message,
          });

          console.warn(
            `Primary model ${primaryModel} failed, falling back to ${fallbackModel}`
          );

          // 重置重试计数，使用回退模型重试
          attempt = -1;
          continue;
        }

        break;
      }
    }

    // 所有尝试都失败
    throw new ModelFallbackError(lastError!, undefined, retryCount);
  }

  /**
   * 记录模型切换事件
   */
  private recordSwitch(event: ModelSwitchEvent): void {
    this.switchHistory.push(event);

    // 保留最近 100 条记录
    if (this.switchHistory.length > 100) {
      this.switchHistory = this.switchHistory.slice(-100);
    }

    // 通知监听器
    for (const listener of this.listeners) {
      try {
        listener(event);
      } catch (e) {
        console.warn('Model switch listener error:', e);
      }
    }
  }

  /**
   * 获取切换历史
   */
  getSwitchHistory(): ModelSwitchEvent[] {
    return [...this.switchHistory];
  }

  /**
   * 添加切换事件监听器
   */
  onSwitch(listener: (event: ModelSwitchEvent) => void): () => void {
    this.listeners.push(listener);
    return () => {
      const index = this.listeners.indexOf(listener);
      if (index !== -1) {
        this.listeners.splice(index, 1);
      }
    };
  }

  /**
   * 清除历史
   */
  clearHistory(): void {
    this.switchHistory = [];
  }

  /**
   * 获取回退统计
   */
  getStats(): {
    totalSwitches: number;
    fallbackRate: number;
    mostRecentSwitch: ModelSwitchEvent | null;
    switchesByModel: Record<string, number>;
  } {
    const switchesByModel: Record<string, number> = {};

    for (const event of this.switchHistory) {
      switchesByModel[event.fromModel] = (switchesByModel[event.fromModel] || 0) + 1;
    }

    return {
      totalSwitches: this.switchHistory.length,
      fallbackRate: this.switchHistory.length > 0
        ? this.switchHistory.filter(e => e.reason === 'fallback').length / this.switchHistory.length
        : 0,
      mostRecentSwitch: this.switchHistory.length > 0
        ? this.switchHistory[this.switchHistory.length - 1]
        : null,
      switchesByModel,
    };
  }
}

/**
 * 全局模型回退实例
 */
export const modelFallback = new ModelFallback();

File: examples/python/models/fallback.py
from __future__ import annotations

import time
from typing import Callable, Iterable, List, Optional

from .config import model_config
from .types import (
    FallbackConfig,
    FallbackCallback,
    ModelFallbackError,
    ModelSwitchEvent,
    RetryCallback,
)


DEFAULT_RETRYABLE_ERRORS = [
    "overloaded_error",
    "rate_limit_error",
    "api_error",
    "timeout",
    "ECONNRESET",
    "ETIMEDOUT",
    "ENOTFOUND",
    "capacity_exceeded",
    "model_unavailable",
]


class ModelFallback:
    def __init__(self, config: Optional[FallbackConfig] = None) -> None:
        resolved_primary = model_config.get_default_model()
        resolved_fallbacks: List[str] = []

        if config:
            resolved_primary = model_config.resolve_alias(config.primaryModel)
            resolved_fallbacks = [model_config.resolve_alias(m) for m in config.fallbackModels]

        self._config = FallbackConfig(
            primaryModel=resolved_primary,
            fallbackModels=resolved_fallbacks,
            retryableErrors=list(config.retryableErrors) if config else list(DEFAULT_RETRYABLE_ERRORS),
            maxRetries=config.maxRetries if config else 3,
            retryDelayMs=config.retryDelayMs if config else 1000,
            exponentialBackoff=config.exponentialBackoff if config else True,
        )
        self._switch_history: List[ModelSwitchEvent] = []
        self._listeners: List[Callable[[ModelSwitchEvent], None]] = []

    def set_primary_model(self, model: str) -> None:
        resolved = model_config.resolve_alias(model)
        if resolved in self._config.fallbackModels:
            raise ValueError("Primary model cannot be the same as a fallback model")
        self._config.primaryModel = resolved

    def set_fallback_models(self, models: Iterable[str]) -> None:
        resolved_models = [model_config.resolve_alias(m) for m in models]
        if self._config.primaryModel in resolved_models:
            raise ValueError("Fallback models cannot include the primary model")
        self._config.fallbackModels = resolved_models

    def add_fallback_model(self, model: str) -> None:
        resolved = model_config.resolve_alias(model)
        if resolved == self._config.primaryModel:
            raise ValueError("Fallback model cannot be the same as primary model")
        if resolved not in self._config.fallbackModels:
            self._config.fallbackModels.append(resolved)

    def get_config(self) -> FallbackConfig:
        return FallbackConfig(
            primaryModel=self._config.primaryModel,
            fallbackModels=list(self._config.fallbackModels),
            retryableErrors=list(self._config.retryableErrors),
            maxRetries=self._config.maxRetries,
            retryDelayMs=self._config.retryDelayMs,
            exponentialBackoff=self._config.exponentialBackoff,
        )

    def is_retryable(self, error: BaseException | object) -> bool:
        if not isinstance(error, BaseException):
            return False
        error_message = str(error).lower()
        error_type = ""
        if hasattr(error, "type"):
            error_type = str(getattr(error, "type")).lower()
        elif hasattr(error, "code"):
            error_type = str(getattr(error, "code")).lower()
        return any(
            retryable.lower() in error_message or retryable.lower() in error_type
            for retryable in self._config.retryableErrors
        )

    def _get_retry_delay(self, attempt: int) -> int:
        if self._config.exponentialBackoff:
            return int(self._config.retryDelayMs * (2 ** attempt))
        return self._config.retryDelayMs

    def _sleep(self, ms: int) -> None:
        time.sleep(ms / 1000)

    def execute_with_fallback(
        self,
        operation: Callable[[str], object],
        *,
        on_retry: Optional[RetryCallback] = None,
        on_fallback: Optional[FallbackCallback] = None,
    ) -> dict:
        models_to_try = [self._config.primaryModel, *self._config.fallbackModels]
        last_error: Optional[Exception] = None
        retries = 0

        for model_index, model in enumerate(models_to_try):
            for attempt in range(self._config.maxRetries + 1):
                try:
                    result = operation(model)
                    return {
                        "result": result,
                        "model": model,
                        "retries": retries,
                        "usedFallback": model_index > 0,
                    }
                except Exception as error:  # noqa: BLE001 - caller handles errors
                    last_error = error
                    retries += 1

                    if self.is_retryable(error) and attempt < self._config.maxRetries:
                        delay = self._get_retry_delay(attempt)
                        if on_retry:
                            on_retry(model, attempt + 1, error)
                        self._sleep(delay)
                        continue

                    break

            if model_index < len(models_to_try) - 1 and last_error and self.is_retryable(last_error):
                next_model = models_to_try[model_index + 1]
                if on_fallback:
                    on_fallback(model, next_model, last_error)
                self._record_switch(
                    ModelSwitchEvent(
                        fromModel=model,
                        toModel=next_model,
                        reason="fallback",
                        timestamp=int(time.time() * 1000),
                        error=str(last_error),
                    )
                )
                continue

            break

        raise ModelFallbackError(last_error or RuntimeError("Unknown error"), model, retries)

    def _record_switch(self, event: ModelSwitchEvent) -> None:
        self._switch_history.append(event)
        if len(self._switch_history) > 100:
            self._switch_history = self._switch_history[-100:]

        for listener in list(self._listeners):
            try:
                listener(event)
            except Exception:
                continue

    def get_switch_history(self) -> List[ModelSwitchEvent]:
        return list(self._switch_history)

    def on_switch(self, listener: Callable[[ModelSwitchEvent], None]) -> Callable[[], None]:
        self._listeners.append(listener)

        def unsubscribe() -> None:
            if listener in self._listeners:
                self._listeners.remove(listener)

        return unsubscribe

    def clear_history(self) -> None:
        self._switch_history = []

    def get_stats(self) -> dict:
        switches_by_model: dict[str, int] = {}
        for event in self._switch_history:
            switches_by_model[event.fromModel] = switches_by_model.get(event.fromModel, 0) + 1

        fallback_events = [event for event in self._switch_history if event.reason == "fallback"]

        return {
            "totalSwitches": len(self._switch_history),
            "fallbackRate": (len(fallback_events) / len(self._switch_history))
            if self._switch_history
            else 0,
            "mostRecentSwitch": self._switch_history[-1] if self._switch_history else None,
            "switchesByModel": switches_by_model,
        }


model_fallback = ModelFallback()
